### PR TITLE
Auto-prover with cyclic notations

### LINKED ISSUE #128: Auto-prover on sequent using recursive notations
@olaure01 suggests:
> I was wondering whether it could be interesting to consider the following approach to cyclic notations: spend 2s searching for a proof without using these cyclic notations at all (just like now), spend 0.5s after a single round of unfolding of each of them, spend 0.5s for a depth 2 unfolding, then fail.

### PR DESCRIPTION
**Pull request by etiennecallies**
Fixes #128

### CODE REVIEW

**Comment by etiennecallies** [comment]
Works on preprod https://linearon.modusponens.dev/?s=o&n=o%2Co-oo
The returned proof is not always the simplest one we could think of https://linearon.modusponens.dev/?s=%3FA&n=A%2C%3FB%2B%3FC&n=B%2CA&n=C%2CA%5E

File: focused_proof.ml [hunk 1]
```diff
@@ -452,16 +433,39 @@ let rec unfocus_proof = function
...
+   and returns the result [(proof opt, is_provable)].
+   If proof option is None, then:
+   * is_provable is false if auto_prover performed an exhaustive research
+   * is_provable is true if max execution time was reached. *)
+let prove_sequent sequent_with_notations =
+    let cyclic_notations, acyclic_notations = split_cyclic_acyclic sequent_with_notations in
+    let max_execution_time_in_seconds = 3. in
+    let ttl = Sys.time () +. max_execution_time_in_seconds in
+    try prove_with_increasing_bound sequent_with_notations.sequent cyclic_notations acyclic_notations [[]] 0 ttl
+    with Ttl_exceeded | Stack_overflow -> (None, true)
```
Comment (etiennecallies) [inline focused_proof.ml:471]: Stack_overflow happens on very simple example like this one https://linearon.modusponens.dev/?s=A&n=A%2CB&n=B%2CA

**Comment by olaure01** [comment]
Fixes #128

File: focused_proof.ml (hunk 1 above)
Comment (olaure01) [reply focused_proof.ml:471]: I do not thing there is any point in unfolding definitions more than, let say, 10 times. Maybe adding such a bound would be nicer than waiting for `Stack_overflow` to be raised.

Comment (etiennecallies) [reply focused_proof.ml]: You're right Stack_overflow is ugly. I changed the way I process in last commit.
If sequents after notation replace don't change (and we don't reach exponential_bound) then we can assert that the research failed after an exhaustive research and therefore mark the sequent with a red turnstile.

**Pull request merged**